daemon/images/images.go

```diff
@@ -202,20 +197,25 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
 					summary.Containers++
 				}
 			}
+			summaryMap[img] = summary
+		}
+
+		summaries = append(summaries, summary)
```

Comment:
There's one thing I don't like about the code; this is not something new, but I've been looking at it when reviewing, and I thought I'd leave a comment about it (haven't thought about the best solution for this yet).

Here, we're constructing:

- a map of summaries, indexed by `img`
- a slice with summaries

I _think_ the reasoning for the map is to de-duplicate multiple images that refer to the same actual "image" (so that we don't re-calculate the size multiple times).

The _slice_ is what we return in the end, but in the code further down (`withExtraAttrs`), we are _indirectly_ updating the slice by updating the map (?); (link to old code, but new code is similar); https://github.com/moby/moby/blob/314759dc2f4745925d8dec6d15acc7761c6e5c92/daemon/images/images.go#L239-L241

I think that's confusing (not directly apparent what's happening); Not sure what a good solution is though.

Reply:
Let's file a separate ticket to look into this, since this PR is severely growing out of scope and it blocks #42531